tests/mir_test_framework/async_server_runner.cpp

```diff
@@ -49,14 +60,36 @@ mtf::AsyncServerRunner::AsyncServerRunner()
     configure_from_commandline(server);
 
     server.add_configuration_option(mtd::logging_opt, mtd::logging_descr, false);
-    server.override_the_logger([&]()
+    server.override_the_logger([&]() -> std::shared_ptr<ml::Logger>
         {
-            std::shared_ptr<ml::Logger> result{};
-
-            if (!server.get_options()->get<bool>(mtd::logging_opt))
-                result = std::make_shared<mtd::NullLogger>();
+            std::ofstream out{output_filename};
+            if (out.good())
+            {
+                std::cerr << "Saving server logs to: " << output_filename << std::endl;
+                if (server.get_options()->get<bool>(mtd::logging_opt))
+                {
+                    return std::make_shared<ml::MultiLogger>(
+                        std::initializer_list<std::shared_ptr<ml::Logger>>{
+                            std::make_shared<ml::DumbConsoleLogger>(),
+                            std::make_shared<ml::FileLogger>(std::move(out))
+                        }
+                    );
+                }
+                else
+                {
+                    return std::make_shared<ml::FileLogger>(std::move(out));
+                }
+            }
+            else
+            {
+                std::cerr << "Failed to open log file: " << output_filename << std::endl;
+                if (!server.get_options()->get<bool>(mtd::logging_opt))
+                {
+                    return std::make_shared<mtd::NullLogger>();
+                }
+            }
 
-            return result;
+            return std::make_shared<ml::DumbConsoleLogger>();
```

Comment:
I know I suggested this, but probably better not to change behaviour:

```suggestion
            // If we return null then the default will be used
            return {};
```